aptos-move/framework/aptos-token/sources/token.move

```diff
@@ -1011,6 +1022,75 @@ module aptos_token::token {
         );
     }
 
+
+    /// Burn a token by creator when the token's BURN_BY_CREATOR is true
+    /// The token is owned at address owner
+    public entry fun burn_by_creator(
+        creator: &signer,
+        owner: address,
+        collection: String,
+        name: String,
+        property_version: u64,
+        amount: u64,
+    ) acquires Collections, TokenStore {
+        let creator_address = signer::address_of(creator);
+        assert!(amount > 0, error::invalid_argument(ENO_BURN_TOKEN_WITH_ZERO_AMOUNT));
+        let token_id = create_token_id_raw(creator_address, collection, name, property_version);
+        let creator_addr = token_id.token_data_id.creator;
+        assert!(
+            exists<Collections>(creator_addr),
+            error::not_found(ECOLLECTIONS_NOT_PUBLISHED),
+        );
+
+        let collections = borrow_global_mut<Collections>(creator_address);
+        assert!(
+            table::contains(&collections.token_data, token_id.token_data_id),
+            error::not_found(ETOKEN_DATA_NOT_PUBLISHED),
+        );
+
+        let token_data = table::borrow_mut(
+            &mut collections.token_data,
+            token_id.token_data_id,
+        );
+
+        // for creator burn token, the property should be explicitly set in the property_map.
+        assert!(
+            property_map::contains_key(&token_data.default_properties,  &string::utf8(BURN_BY_CREATOR)),
+            error::permission_denied(ETOKEN_CANNOT_BURN_BY_CREATOR)
+        );
+
+        let burn_by_owner_flag = property_map::read_bool(&token_data.default_properties, &string::utf8(BURN_BY_CREATOR));
+        assert!(burn_by_owner_flag, error::permission_denied(ETOKEN_CANNOT_BURN_BY_CREATOR));
+
+        // Burn the tokens.
+        let Token { id: _, amount: burned_amount, token_properties: _ } = withdraw_with_event_internal(owner, token_id, amount);
+        let token_store = borrow_global_mut<TokenStore>(owner);
+        event::emit_event<BurnTokenEvent>(
+            &mut token_store.burn_events,
+            BurnTokenEvent { id: token_id, amount: burned_amount},
+        );
+
+        // Decrease the supply correspondingly by the amount of tokens burned.
+        token_data.supply = token_data.supply - burned_amount;
+
+        // Delete the token_data if supply drops to 0.
+        if (token_data.supply == 0) {
```

Comment:
Wasn't there a bug related to this earlier where max = 0 means the token supply is not tracked? Shouldn't we check the max here as well?

Reply:
rebased when the other fix PR landed